### ISSUE TITLE
Skip warning about CTEs when SETTINGS submitted

### ISSUE THREAD

**Issue by rustprooflabs** [pgMustard/pgMustard#48]
Would it be possible to omit the warning about CTEs when the plan was also submitted with details from `SETTINGS`?  This option was added to `EXPLAIN` in Pg12, the same version that resolves the issue being warned about.

> "In versions of Postgres prior to v12, CTEs (caused by using "WITH...AS..." structures) can prevent..."

Example: https://app.pgmustard.com/#/explore/af143df0-2863-43d4-9ac0-adda7c741063

**Comment by hairybreeches** [comment]
Hi, and thanks for the suggestion!

I'm really open to doing this kind of thing: there are quite a few version hints in the plans, and we've talked a lot about using those to provide better, more targeted advice based on version.

However, in this case, I'm not sure it's the best option.

The patch notes detailing the new process of inlining CTEs says:

> Allow common table expressions (CTEs) to be inlined into the outer query (Andreas Karlsson, Andrew Gierth, David Fetter, Tom Lane)
> Specifically, CTEs are automatically inlined if they have no side-effects, are not recursive, and are referenced only once in the query. Inlining can be prevented by specifying MATERIALIZED, or forced for multiply-referenced CTEs by specifying NOT MATERIALIZED. Previously, CTEs were never inlined and were always evaluated before the rest of the query.

I think that still leaves quite a few cases where inlining a CTE might not be done automatically, but is still be possible and has the potential to yield some performance improvements. For example your plan might be one where, although it has not been inlined automatically, it could be possible to do so manually (although I doubt in this case any significant performance gain would result).

I think instead I might suggest altering the description here to place much less emphasis on the auto-inlining functionality introduced in v12, since I do believe in many cases manual inlining could provide value to people using postgres 12+. This was not something we expected when this text was written (iirc at the same time v12 was coming out), but I think examples like this prove us wrong.

However this also brings up another point - I think that we are in general over-scoring these CTE tips in cases like this, where it seems unlikely that inlining will allow postgres to provide performance improvement (for example, there are no filters or sorts discarding the bulk of the rows produced by the CTE). I think it would be really valuable for us to spend some time reducing this over-scoring, since this tip does appear too often (I think yours is a perfect example).

In summary:

- I now believe that this suggestion still has value for people using v12
- I suggest instead we update the text of the tip to significantly de-emphasise auto-inlining
- I think we should spend time working on the scoring of this tip since it is frequently over-scored.

What do you reckon?

**Comment by rustprooflabs** [comment]
Those are great points.  It seems that when Pg inlines the CTE query the details about being a CTE are removed from the query plan. Thus I (the user) wouldn't be warned about it.  When the plan still includes details about the CTE it seems that the user should be informed.

I had to spend some time looking at the different scenarios to think about scoring, I used this table with 1M rows for some examples.

```sql
DROP TABLE IF EXISTS x;
CREATE TABLE x AS
SELECT id, md5(id::TEXT) AS val
	FROM generate_series(1,1000000) id
;
ALTER TABLE x ADD CONSTRAINT pk_x PRIMARY KEY (id);
```

An easy query for demonstrating inlined CTEs in Pg12 and later (tested Pg13).

```sql
EXPLAIN (ANALYZE, COSTS OFF)
WITH a AS (
SELECT * FROM x WHERE id = 499999
)
SELECT * FROM a
;
```

On Pg 11 this returns the plan:

```bash
                                  QUERY PLAN                                  
------------------------------------------------------------------------------
 CTE Scan on a (actual time=0.027..0.028 rows=1 loops=1)
   CTE a
     ->  Index Scan using pk_x on x (actual time=0.024..0.025 rows=1 loops=1)
           Index Cond: (id = 499999)
 Planning Time: 0.104 ms
 Execution Time: 0.051 ms
```

Postgres 13 removes the CTE Scan, so this query doesn't generate a warning about the CTE in the app. (I double checked!)

```bash
                              QUERY PLAN                              
----------------------------------------------------------------------
 Index Scan using pk_x on x (actual time=0.031..0.031 rows=1 loops=1)
   Index Cond: (id = 499999)
 Planning Time: 0.277 ms
 Execution Time: 0.052 ms
```

A change to the query to reference the CTE `a` twice gets Pg13 to materialize, thus the CTE Scan is reported.  In this case, the fact it materialized may or may not be a good thing for performance.  I wouldn't expect it to be ranked a 5 in the app, though it's definitely worth mentioning the reasons why a CTE might materialize and consider reviewing those.  The query I originally submitted is in this category of "there's probably something worth looking at here."

```sql
EXPLAIN (ANALYZE, COSTS OFF)
WITH a AS (
SELECT * FROM x WHERE id > 450000 AND id < 500000
), b AS (
SELECT AVG(id) AS avg FROM a
)
SELECT * FROM a
;
```

```bash
                                    QUERY PLAN                                     
-----------------------------------------------------------------------------------
 CTE Scan on a (actual time=0.030..25.429 rows=49999 loops=1)
   CTE a
     ->  Index Scan using pk_x on x (actual time=0.028..13.729 rows=49999 loops=1)
           Index Cond: ((id > 450000) AND (id < 500000))
 Planning Time: 0.284 ms
 Execution Time: 29.888 ms
```

In the above case, a suggestion could be made to add `NOT MATERIALIZED`, but this might not always be the best either. In this case it's faster.

```sql
EXPLAIN (ANALYZE, COSTS OFF)
WITH a AS NOT MATERIALIZED (
SELECT * FROM x WHERE id > 450000 AND id < 500000
), b AS (
SELECT AVG(id) AS avg FROM a
)
SELECT * FROM a
;
```

```bash
                                QUERY PLAN                                 
---------------------------------------------------------------------------
 Index Scan using pk_x on x (actual time=0.052..10.970 rows=49999 loops=1)
   Index Cond: ((id > 450000) AND (id < 500000))
 Planning Time: 0.137 ms
 Execution Time: 13.594 ms
```

Recursive CTEs are a topic of their own.  My gut says that when a recursive CTE is being used, complex logic is in play and a better option likely does not exist.  Or, of course, they could just be over complicating the logic and just think they need a recursive CTE.  I've been on both sides of that fence! :smile:

**Comment by hairybreeches** [comment]
That's all really interesting data, really appreciate the result of experiments like these!

Given what you've said, I've removed the references to inlining in v12+, so the tip now reads:

> CTEs (caused by using "WITH...AS..." structures) can prevent query plan optimisations, since the planner isn't aware of where and how the result of the CTE will be used.
> 
> This can lead to suboptimal plans for CTEs, particularly those returning results which are only referred to once in the rest of the query.
> 
> Inlining such a CTE, by replacing references with the code used to generate the value, can allow Postgres to better optimise your query, leading to surprising gains in performance.

I think we already deal with CTEs used multiple times pretty well - the scoring is such that they are unlikely to show up very high on the list.

I do think there's more we can do to improve the scoring of this tip though, and improving that logic is something I hope to spend some time on over the next few weeks.

**Comment by mchristofides** [comment]
Sorry for the ping on a super old thread, but just wanted to update this to say I've shipped some changes to this tip, including the wording, scoring, and the "Learn more" link.

The tip text now reads:
> CTEs (caused by using "WITH...AS..." structures) can hurt performance, mostly depending on whether or not Postgres materializes the result.
>
>You can test by forcing materialization on and off. Inlining subqueries can also still be beneficial for CTEs which are only referred to once in the rest of the query.

And the learn more link now points to: https://boringsql.com/posts/good-cte-bad-cte

Thanks again the suggestions and discussion!
